### PR TITLE
1972 how are we going to merge spatial structure

### PR DESCRIPTION
**Pull request by msevestre**

### CODE REVIEW

File: tests/OSPSuite.HelpersForTests/ModuleHelperForSpecs.cs [hunk 1]
```diff
@@ -158,10 +160,82 @@ private IParameter newConstantParameter(string name, double value, ParameterBuil
             .WithMode(parameterBuildMode);
       }
 
+      //Module 2 will introduce
```
Comment (msevestre) [inline tests/OSPSuite.HelpersForTests/ModuleHelperForSpecs.cs:163]: @Yuri05 @rwmcintosh @PavelBal  Interested in added simple use cases here
Can we validate that this is the expected behavior?

File: tests/OSPSuite.HelpersForTests/ModuleHelperForSpecs.cs [hunk 2]
```diff
@@ -158,10 +160,82 @@ private IParameter newConstantParameter(string name, double value, ParameterBuil
...
+         var neighborhood2 = _neighborhoodFactory.CreateBetween(lngPlasma, heartPlasma).WithName("lng_pls_to_hrt_pls");
+         neighborhood2.AddParameter(newConstantParameter("SA", 10));
+         spatialStructure.AddNeighborhood(neighborhood2);
+
+         module.SpatialStructure = spatialStructure;
+         return module;
+      }
+
+      //Module 3 will add a tumor organ in the heart organ created above
+      private Module createModule3()
```
Comment (msevestre) [inline tests/OSPSuite.HelpersForTests/ModuleHelperForSpecs.cs:219]: @Yuri05 @rwmcintosh @PavelBal  Interested in added simple use cases here

**Review by msevestre — COMMENTED** [review]
This is a draft but questions to be answered please

File: src/OSPSuite.Core/Domain/Mappers/NeighborhoodCollectionToContainerMapper.cs [hunk 3]
```diff
@@ -44,11 +44,13 @@ public IContainer MapFrom(ModelConfiguration modelConfiguration)
 
          var moleculeNamesCopyProperties = simulationBuilder.AllPresentXenobioticFloatingMoleculeNames();
 
-         simulationBuilder.SpatialStructures.SelectMany(x => x.Neighborhoods).Each(nb =>
-            neighborhoodsParentContainer.Add(_neighborhoodMapper.MapFrom(nb,
-               moleculeNamesFor(nb, startValuesForFloatingMolecules),
-               moleculeNamesCopyProperties, modelConfiguration)));
+         //we use a cache to ensure that we are replacing neighborhoods defined in multiple structures
+         var neighborhoodCache = new ObjectBaseCache<Neighborhood>();
```
Comment (msevestre) [inline src/OSPSuite.Core/Domain/Mappers/NeighborhoodCollectionToContainerMapper.cs:48]: Usage of cache here by name to ensure that the last neighborhood wins

File: src/OSPSuite.Core/Domain/Services/SpatialStructureMerger.cs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            .WithContainerType(ContainerType.Simulation);
+
+         root.AddTag(new Tag(Constants.ROOT_CONTAINER_TAG));
+
+         createMergedContainerStructureInRoot(root, simulationBuilder);
+
+         return root;
+      }
+
+      private void createMergedContainerStructureInRoot(IContainer root, SimulationBuilder simulationBuilder)
```
Comment (msevestre) [inline src/OSPSuite.Core/Domain/Services/SpatialStructureMerger.cs:50]: @rwmcintosh @Yuri05 Please review this. only a few lines of code dealing with merging, replacing etc...I think this is what we want

**Review by rwmcintosh — APPROVED** [review]

**Comment by msevestre** [comment]
@Yuri05 Please review the code

**Pull request merged**